infer/inference.cc

```diff
@@ -128,17 +128,61 @@ unique_ptr<cfg::CFG> Inference::run(core::Context ctx, unique_ptr<cfg::CFG> cfg)
             bb->firstDeadInstructionIdx = 0;
             // this block is unreachable.
             if (!bb->exprs.empty()) {
+                // This is a bit complicated:
+                //
+                //   1. If the block consists only of synthetic bindings or T.absurd, we don't
+                //      want to issue an error.
+                //   2. If the block contains a send of the form <Magic>.<nil-for-safe-navigation>(x),
+                //      we want to issue an UnnecessarySafeNavigationError, extracting
+                //      type-and-origin info from x. (This magic form is inserted by the desugarer
+                //      for a "safe navigation" operation, e.g., `x&.foo`.)
+                //   3. Otherwise, we want to issue a DeadBranchInferencer error, taking the first
+                //      (non-synthetic, non-"T.absurd") instruction in the block as the loc of the
+                //      error.
+                cfg::Instruction *unreachableInstruction = nullptr;
+                core::LocOffsets locForUnreachable;
+                bool dueToSafeNavigation = false;
+
                 for (auto &expr : bb->exprs) {
                     if (expr.value->isSynthetic) {
                         continue;
                     }
                     if (cfg::isa_instruction<cfg::TAbsurd>(expr.value.get())) {
                         continue;
                     }
-                    if (auto e = ctx.beginError(expr.loc, core::errors::Infer::DeadBranchInferencer)) {
+
+                    auto send = cfg::cast_instruction<cfg::Send>(expr.value.get());
+                    if (send != nullptr && send->fun == core::Names::nilForSafeNavigation()) {
+                        unreachableInstruction = expr.value.get();
```

Comment:
Why not emit the error right here and break? Why have a separate variable that we use to emit the error later?

Reply:
If we find the send to `nilForSafeNavigation` we could go ahead and emit the error and break. The tricky thing is that for other instructions we can't emit the ordinary "code is unreachable" error right away, because the magic send unfortunately does not wind up occurring first. The guts of the dead block wind up looking like this, with the magic send in the second position:

```
<cfgAlias>$28: T.class_of(<Magic>) = alias <C <Magic>>
<statTemp>$21: NilClass = <cfgAlias>$28: T.class_of(<Magic>).<nil-for-safe-navigation>(y: NilClass)
<unconditional>
```

When you put it all together, doing it this way ended up feeling a little cleaner to me.

Reply:
Gotcha. I guess i was just surprised to see how much code this required but it seems to get the job done nonetheless so 👍

Reply:
> surprised to see how much code this required

same tbh  ;)